design-proposals/vm-delete-protection.md

```diff
@@ -0,0 +1,155 @@
+by KubeVirt. By default, this annotation is not present on a VM. It is up to the user to enable the deletion protection.
+The annotation looks like this in the VM manifests:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  annotations:
+    kubevirt.io/vm-delete-protection: "true"
+spec:
+[...]
+```
+
+### Backend
+
+#### Option \#1: ValidatingAdmissionPolicy
+
+The VM protection is implemented by adding
+a  [ValidationAdmissionPolicy and ValidationAdmissionPolicyBinding](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/)
+objects. On every delete action against a VM object, the ValidatingAdmissionPolicy will check for presence of the
+annotation. If it is present and its value equals “true” or “True” the request will be rejected. The
+ValidationAdmissionPolicy and ValidationAdmissionPolicyBinding objects will be configured with the following parameters:
+
+- matchConstrains: Apply to \`virtualmachine\` objects; Operations: DELETE.
+- Variables:
+    - \`annotation\`: Returns the VM annotation field;
+        - string('kubevirt.io/vm-delete-protection')
+- Validation rules:
+    - Checks if the annotation is present at all, if present, checks if it is equal to “true” or “True”;
+        - (\!(variables.annotation in oldObject.metadata.annotations) || \!oldObject.metadata.annotations\[variables.annotation\].matches('^(true|True)$'))
+
+Check [https://github.com/kubevirt/community/blob/main/design-proposals/shadow-node.md\#backwards-compatibility](https://github.com/kubevirt/community/blob/main/design-proposals/shadow-node.md#backwards-compatibility)
+for backwards compatibility details.
+
+Pros:
+
+- Kubernetes built-in feature which is straightforward to maintain, implement and configure.
+- ValidationAdmisssionPolicy is [GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) since Kuberentes version 1.30.
+
+Cons:
+
+- It will not work in deployments using Kubernetes versions \< 1.30
+
+A proof of concept implementation can be found
+at: [https://github.com/kubevirt/kubevirt/commit/7167e68723201b3899f8f36b02e5bf11701ee722](https://github.com/kubevirt/kubevirt/commit/7167e68723201b3899f8f36b02e5bf11701ee722)
+
+#### Option 2\#: Virtual Machine Admission Webhook
+
+A new validation rule is added to the VM admitters. The admitter will check for presence of
+the \`[kubevirt.io/vm-delete-protection](http://kubevirt.io/vm-delete-protection)\` annotation, filter delete
+operations, and will reject any attempt to delete VM objects if the annotation is present and its value is set to “true”
+or “True”.
+
+Pros:
+
+- It works with any Kubernetes version.
+
+Cons:
+
+- It will increase the maintenance effort.
+
+A proof of concept implementation can be found
+at:  [https://github.com/kubevirt/kubevirt/commit/2fafcf649019c26c8802838f2bbe673a9e93f04c](https://github.com/kubevirt/kubevirt/commit/2fafcf649019c26c8802838f2bbe673a9e93f04c)
+
+#### Option 3\#: Finalizers
```

Comment:
After understanding the requested requirements, I don't think we should consider this as an option. Maybe just add a note about it that isn't releavnt to us.

Reply:
The discussion about finalizers has been triggered in multiple forums. Therefore, I've considered it relevant to add a specific Section explaining why we cannot implement this feature thought finalizers.